### PR TITLE
Improve ISD performance (Part two)

### PR DESCRIPTION
**Pull request by palemieux**

### CODE REVIEW

File: src/main/python/ttconv/isd.py [hunk 1]
```diff
@@ -36,6 +36,23 @@
 import ttconv.model as model
 import ttconv.style_properties as styles
 
+class SignificantTimes:
+  def __init__(self, sig_times, cache = None):
```
Comment (valnoel) [inline src/main/python/ttconv/isd.py:40]: Could you add type hints here, please? 🙏

**Review by valnoel — CHANGES_REQUESTED** [review]

File: src/main/python/ttconv/srt/writer.py [hunk 2]
```diff
@@ -174,26 +174,24 @@ def __str__(self) -> str:
...
 
-  for offset, isd in isds:
+  for i, (offset, isd) in enumerate(isds):
 
     for srt_filter in srt.filters:
       srt_filter.process(isd)
 
     srt.add_isd(isd, offset)
 
-    progress_callback((index + 1) / nb_significant_times)
```
Comment (valnoel) [inline src/main/python/ttconv/srt/writer.py:191]: It means that the "writing" progress is only defined by the ISD generation? Shouldn't we share the percentage of progress between ISD and SRT generations (70/30 for instance)?

**Review by valnoel — CHANGES_REQUESTED** [review]

File: src/main/python/ttconv/srt/writer.py (hunk 2 above)
Comment (palemieux) [reply src/main/python/ttconv/srt/writer.py:191]: Done at https://github.com/sandflow/ttconv/pull/160/commits/ced81f604b69004903af94b587da33ce60094c35

File: src/main/python/ttconv/isd.py [hunk 3]
```diff
@@ -190,31 +202,70 @@ def sig_times(element: model.ContentElement, parent_begin: Fraction, parent_end:
...
+  @staticmethod
+  def generate_isd_sequence(doc: model.ContentDocument, is_multithreaded: bool = True) -> typing.List[typing.Tuple[Fraction, ISD]]:
+    """ Returns a list of duples, each consisting of a significant time in the ContentDocument `doc`
+    and the corresponding `ISD` instance. The duples are sorted in order of increasing significant time.
+    This helper function by default uses multithreading to improve performance on large ContentDocument instances.
+    """
+
+    sig_times = ISD.significant_times(doc)
+
+  # Compute ISDs
```
[the remark this answers is not in the document]
Comment (palemieux) [reply src/main/python/ttconv/isd.py]: Fixed at https://github.com/sandflow/ttconv/pull/160/commits/0ca6b5f1c04a34a29f39f558f05f29a4b053f554

File: src/main/python/ttconv/isd.py (hunk 1 above)
Comment (palemieux) [reply src/main/python/ttconv/isd.py]: Fixed

**Review by valnoel — APPROVED** [review]

**Pull request merged**